src/python/pants/backend/go/goals/generate.py

```diff
@@ -0,0 +1,285 @@
+            "GOLINE": str(line_num),
+            "GOPACKAGE": analysis.name,
+            "GOROOT": goroot.path,
+            "DOLLAR": "$",
+        }
+
+        for i, arg in enumerate(args):
+            args[i] = _expand_env(arg, generate_env)
+
+        # Invoke the subprocess and store its output for use as input root of next command (if any).
+        result = await Get(
+            ProcessResult,
+            Process(
+                argv=args,
+                input_digest=digest,
+                working_directory=dir_path,
+                output_directories=["."],
+                env={
+                    **generate_env,
+                    **env,
+                },
+                description=f"Process `go generate` directives in file: {os.path.join(dir_path, go_file)}",
+            ),
+        )
+        digest = await Get(Digest, AddPrefix(result.output_digest, dir_path))
+
+    return digest
+
+
+@rule
+async def run_go_package_generators(
+    request: RunPackageGeneratorsRequest, goroot: GoRoot, subsystem: GoGenerateGoalSubsystem
+) -> RunPackageGeneratorsResult:
+    fallible_analysis, env = await MultiGet(
+        Get(
+            FallibleFirstPartyPkgAnalysis,
+            FirstPartyPkgAnalysisRequest(request.address, extra_build_tags=("generate",)),
+        ),
+        Get(EnvironmentVars, EnvironmentVarsRequest(subsystem.env_vars)),
+    )
+    if not fallible_analysis.analysis:
+        raise ValueError(f"Analysis failure for {request.address}: {fallible_analysis.stderr}")
+    analysis = fallible_analysis.analysis
+    dir_path = analysis.dir_path if analysis.dir_path else "."
+
+    fallible_pkg_digest = await Get(
+        FallibleFirstPartyPkgDigest, FirstPartyPkgDigestRequest(request.address)
+    )
+    if fallible_pkg_digest.pkg_digest is None:
+        raise ValueError(
+            f"Unable to obtain digest for {request.address}: {fallible_pkg_digest.stderr}"
+        )
+    pkg_digest = fallible_pkg_digest.pkg_digest
+
+    # Scan each Go file in the package for generate directives. Process them sequentially so that an error in
+    # an earlier-processed file prevents later files from being processed.
+    output_digest = EMPTY_DIGEST
+    for go_file in analysis.go_files:
+        output_digest_for_go_file = await _run_generators(
+            analysis, pkg_digest.digest, dir_path, go_file, goroot, env
+        )
+        output_digest = await Get(Digest, MergeDigests([output_digest, output_digest_for_go_file]))
+
+    return RunPackageGeneratorsResult(output_digest)
+
+
+@goal_rule
+async def go_generate(targets: Targets, workspace: Workspace) -> GoGenerateGoal:
+    go_package_targets = [tgt for tgt in targets if tgt.has_field(GoPackageSourcesField)]
+    results = await MultiGet(
+        Get(RunPackageGeneratorsResult, RunPackageGeneratorsRequest(tgt.address))
+        for tgt in go_package_targets
+    )
+    output_digest = await Get(Digest, MergeDigests([r.digest for r in results]))
```

Comment:
My understanding is that the output directory is `.` each time? Do we expect issues with merge conflicts?

Reply:
That I don't know. The rules cannot know what files the generators will actually output so it is probably on users to not write the same file with different content. At the very least, Pants would error in a case where `go generate` would just silently allow the generators to overwrite each other's output.

Reply:
The error message is pretty wonky for merge conflicts and will confuse most people. Thoughts on proactively scanning for merge conflicts and eagerly erroring with an informative error message? I think you'd want to check both inside the rule for an individual target, and then also here.

Reply:
Is there a helper to do that?

Reply:
`SnapshotDiff.from_snapshots(self.input, self.output)`

Reply:
Merging was actually wrong here. The "new" digest should overwrite files from the "original" digest. To that end, I've added a helper function to merge the digests in that manner.